File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax/SingleLineLambdaExpressionSyntax.cs
using System;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
{
	internal sealed class SingleLineLambdaExpressionSyntax : LambdaExpressionSyntax
	{
		internal readonly VisualBasicSyntaxNode _body;

		internal static Func<ObjectReader, object> CreateInstance;

		internal VisualBasicSyntaxNode Body => _body;

		internal SingleLineLambdaExpressionSyntax(SyntaxKind kind, LambdaHeaderSyntax subOrFunctionHeader, VisualBasicSyntaxNode body)
			: base(kind, subOrFunctionHeader)
		{
			base._slotCount = 2;
			AdjustFlagsAndWidth(body);
			_body = body;
		}

		internal SingleLineLambdaExpressionSyntax(SyntaxKind kind, LambdaHeaderSyntax subOrFunctionHeader, VisualBasicSyntaxNode body, ISyntaxFactoryContext context)
			: base(kind, subOrFunctionHeader)
		{
			base._slotCount = 2;
			SetFactoryContext(context);
			AdjustFlagsAndWidth(body);
			_body = body;
		}

		internal SingleLineLambdaExpressionSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, LambdaHeaderSyntax subOrFunctionHeader, VisualBasicSyntaxNode body)
			: base(kind, errors, annotations, subOrFunctionHeader)
		{
			base._slotCount = 2;
			AdjustFlagsAndWidth(body);
			_body = body;
		}

		internal SingleLineLambdaExpressionSyntax(ObjectReader reader)
			: base(reader)
		{
			base._slotCount = 2;
			VisualBasicSyntaxNode visualBasicSyntaxNode = (VisualBasicSyntaxNode)reader.ReadValue();
			if (visualBasicSyntaxNode != null)
			{
				AdjustFlagsAndWidth(visualBasicSyntaxNode);
				_body = visualBasicSyntaxNode;
			}
		}

		internal override void WriteTo(ObjectWriter writer)
		{
			base.WriteTo(writer);
			writer.WriteValue(_body);
		}

		static SingleLineLambdaExpressionSyntax()
		{
			CreateInstance = (ObjectReader o) => new SingleLineLambdaExpressionSyntax(o);
			ObjectBinder.RegisterTypeReader(typeof(SingleLineLambdaExpressionSyntax), (ObjectReader r) => new SingleLineLambdaExpressionSyntax(r));
		}

		internal override SyntaxNode CreateRed(SyntaxNode parent, int startLocation)
		{
			return new Microsoft.CodeAnalysis.VisualBasic.Syntax.SingleLineLambdaExpressionSyntax(this, parent, startLocation);
		}

		internal override GreenNode GetSlot(int i)
		{
			return i switch
			{
				0 => _subOrFunctionHeader, 
				1 => _body, 
				_ => null, 
			};
		}

		internal override GreenNode SetDiagnostics(DiagnosticInfo[] newErrors)
		{
			return new SingleLineLambdaExpressionSyntax(base.Kind, newErrors, GetAnnotations(), _subOrFunctionHeader, _body);
		}

		internal override GreenNode SetAnnotations(SyntaxAnnotation[] annotations)
		{
			return new SingleLineLambdaExpressionSyntax(base.Kind, GetDiagnostics(), annotations, _subOrFunctionHeader, _body);
		}

		public override VisualBasicSyntaxNode Accept(VisualBasicSyntaxVisitor visitor)
		{
			return visitor.VisitSingleLineLambdaExpression(this);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/InternalSyntax/SingleLineLambdaExpressionSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Roslyn.Utilities
Imports System
Imports System.Runtime.CompilerServices

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
	Friend NotInheritable Class SingleLineLambdaExpressionSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.LambdaExpressionSyntax
		Friend ReadOnly _body As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode

		Friend Shared CreateInstance As Func(Of ObjectReader, Object)

		Friend ReadOnly Property Body As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode
			Get
				Return Me._body
			End Get
		End Property

		Shared Sub New()
			Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.SingleLineLambdaExpressionSyntax.CreateInstance = Function(o As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.SingleLineLambdaExpressionSyntax(o)
			ObjectBinder.RegisterTypeReader(GetType(Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.SingleLineLambdaExpressionSyntax), Function(r As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.SingleLineLambdaExpressionSyntax(r))
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal subOrFunctionHeader As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.LambdaHeaderSyntax, ByVal body As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode)
			MyBase.New(kind, subOrFunctionHeader)
			MyBase._slotCount = 2
			MyBase.AdjustFlagsAndWidth(body)
			Me._body = body
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal subOrFunctionHeader As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.LambdaHeaderSyntax, ByVal body As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode, ByVal context As ISyntaxFactoryContext)
			MyBase.New(kind, subOrFunctionHeader)
			MyBase._slotCount = 2
			MyBase.SetFactoryContext(context)
			MyBase.AdjustFlagsAndWidth(body)
			Me._body = body
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal subOrFunctionHeader As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.LambdaHeaderSyntax, ByVal body As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode)
			MyBase.New(kind, errors, annotations, subOrFunctionHeader)
			MyBase._slotCount = 2
			MyBase.AdjustFlagsAndWidth(body)
			Me._body = body
		End Sub

		Friend Sub New(ByVal reader As ObjectReader)
			MyBase.New(reader)
			MyBase._slotCount = 2
			Dim visualBasicSyntaxNode As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode)
			If (visualBasicSyntaxNode IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(visualBasicSyntaxNode)
				Me._body = visualBasicSyntaxNode
			End If
		End Sub

		Public Overrides Function Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxVisitor) As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode
			Return visitor.VisitSingleLineLambdaExpression(Me)
		End Function

		Friend Overrides Function CreateRed(ByVal parent As SyntaxNode, ByVal startLocation As Integer) As SyntaxNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.SingleLineLambdaExpressionSyntax(Me, parent, startLocation)
		End Function

		Friend Overrides Function GetSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.GreenNode
			Dim greenNode As Microsoft.CodeAnalysis.GreenNode
			Dim num As Integer = i
			If (num = 0) Then
				greenNode = Me._subOrFunctionHeader
			ElseIf (num = 1) Then
				greenNode = Me._body
			Else
				greenNode = Nothing
			End If
			Return greenNode
		End Function

		Friend Overrides Function SetAnnotations(ByVal annotations As SyntaxAnnotation()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.SingleLineLambdaExpressionSyntax(MyBase.Kind, MyBase.GetDiagnostics(), annotations, Me._subOrFunctionHeader, Me._body)
		End Function

		Friend Overrides Function SetDiagnostics(ByVal newErrors As DiagnosticInfo()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.SingleLineLambdaExpressionSyntax(MyBase.Kind, newErrors, MyBase.GetAnnotations(), Me._subOrFunctionHeader, Me._body)
		End Function

		Friend Overrides Sub WriteTo(ByVal writer As ObjectWriter)
			MyBase.WriteTo(writer)
			writer.WriteValue(DirectCast(Me._body, IObjectWritable))
		End Sub
	End Class
End Namespace